File: packages/protocol-schema/src/models/device.ts
import { Type, type Static } from "@sinclair/typebox";

export const DevicePlatformSchema = Type.Union(
  [Type.Literal("ios"), Type.Literal("desktop")],
  { $id: "DevicePlatform" }
);
export type DevicePlatform = Static<typeof DevicePlatformSchema>;

export const DeviceTransportSchema = Type.Union(
  [Type.Literal("lan"), Type.Literal("tailscale")],
  { $id: "DeviceTransport" }
);
export type DeviceTransport = Static<typeof DeviceTransportSchema>;

export const DeviceSchema = Type.Object(
  {
    id: Type.String(),
    name: Type.String(),
    platform: DevicePlatformSchema,
    publicKey: Type.String(),
    pairedAt: Type.Integer(),
    lastSeenAt: Type.Integer(),
    lastTransport: DeviceTransportSchema,
    revoked: Type.Boolean(),
  },
  { $id: "Device" }
);
export type Device = Static<typeof DeviceSchema>;

export const PairingQrPayloadSchema = Type.Object(
  {
    hostId: Type.String(),
    hostName: Type.String(),
    token: Type.String(),
    endpoints: Type.Array(Type.String()),
    hostPublicKey: Type.String(),
    expiresAt: Type.Integer(),
  },
  { $id: "PairingQrPayload" }
);
export type PairingQrPayload = Static<typeof PairingQrPayloadSchema>;

export const PairingRequestSchema = Type.Object(
  {
    token: Type.String(),
    deviceId: Type.String(),
    deviceName: Type.String(),
    devicePublicKey: Type.String(),
    platform: DevicePlatformSchema,
  },
  { $id: "PairingRequest" }
);
export type PairingRequest = Static<typeof PairingRequestSchema>;

export const PairingResponseSchema = Type.Object(
  {
    status: Type.Literal("paired"),
    hostId: Type.String(),
    authToken: Type.String(),
  },
  { $id: "PairingResponse" }
);
export type PairingResponse = Static<typeof PairingResponseSchema>;

export const AuthChallengeSchema = Type.Object(
  {
    nonce: Type.String(),
    hostId: Type.String(),
    timestamp: Type.Integer(),
  },
  { $id: "AuthChallenge" }
);
export type AuthChallenge = Static<typeof AuthChallengeSchema>;

export const AuthSolutionSchema = Type.Object(
  {
    deviceId: Type.String(),
    signature: Type.String(),
  },
  { $id: "AuthSolution" }
);
export type AuthSolution = Static<typeof AuthSolutionSchema>;

File: crates/canywhere-protocol/src/models/device.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub enum DevicePlatform {
    Ios,
    Desktop,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub enum DeviceTransport {
    Lan,
    Tailscale,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct Device {
    pub id: String,
    pub name: String,
    pub platform: DevicePlatform,
    pub public_key: String,
    pub paired_at: i64,
    pub last_seen_at: i64,
    pub last_transport: DeviceTransport,
    pub revoked: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct PairingQrPayload {
    pub host_id: String,
    pub host_name: String,
    pub token: String,
    pub endpoints: Vec<String>,
    pub host_public_key: String,
    pub expires_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct PairingRequest {
    pub token: String,
    pub device_id: String,
    pub device_name: String,
    pub device_public_key: String,
    pub platform: DevicePlatform,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
pub struct PairingResponse {
    pub status: String, // "paired"
    pub host_id: String,
    pub auth_token: String,
}
